contrib/vcl/source/vcl_interface.cc

```diff
@@ -0,0 +1,136 @@
+  int wrk_index = vppcom_worker_index();
+  VCL_LOG("events on worker {}", wrk_index);
+  struct epoll_event events[MaxNumEpollEvents];
+  int max_events = MaxNumEpollEvents;
+  uint32_t epoll_fd = vclEpollHandle(wrk_index);
+
+  while (max_events > 0) {
+    int n_events = vppcom_epoll_wait(epoll_fd, events, max_events, 0);
+    if (n_events <= 0) {
+      break;
+    }
+    max_events -= n_events;
+    VCL_LOG("had {} events", n_events);
+
+    for (int i = 0; i < n_events; i++) {
+      auto vcl_handle = reinterpret_cast<VclIoHandle*>(events[i].data.u64);
+      if (vcl_handle->isWrkListener()) {
+        vcl_handle = vcl_handle->getParentListener();
+      }
+
+      // session closed due to some recently processed event
+      if (!vcl_handle->isOpen()) {
+        continue;
+      }
+
+      uint32_t evts = 0;
+      if (events[i].events & EPOLLIN) {
+        evts |= Event::FileReadyType::Read;
+      }
+      if (events[i].events & EPOLLOUT) {
+        evts |= Event::FileReadyType::Write;
+      }
+      if (events[i].events & (EPOLLERR | EPOLLHUP)) {
+        evts |= Event::FileReadyType::Closed;
+      }
+
+      VCL_LOG("got event on vcl handle fd {} sh {:x} events {}", vcl_handle->fdDoNotUse(),
+              vcl_handle->sh(), evts);
+      vcl_handle->cb(evts);
+    }
+  }
+}
+
+} // namespace
+
+uint32_t vclEpollHandle(uint32_t wrk_index) {
+  std::vector<uint32_t>& epoll_handles = epollHandles();
+  return epoll_handles[wrk_index];
+}
+
+void vclInterfaceWorkerRegister() {
+  {
+    absl::MutexLock lk(&wrk_lock);
+    RELEASE_ASSERT(vppcom_worker_register() == VPPCOM_OK, "failed to register VCL worker");
+  }
+  const int wrk_index = vppcom_worker_index();
+  int epoll_handle = vppcom_epoll_create();
+  std::vector<uint32_t>& epoll_handles = epollHandles();
+  epoll_handles[wrk_index] = epoll_handle;
+  VCL_LOG("registered worker {} and epoll handle {:x} mq fd {}", wrk_index, epoll_handle,
+          vppcom_mq_epoll_fd());
+}
+
+void vclInterfaceRegisterEpollEvent(Envoy::Event::Dispatcher& dispatcher) {
+  MqFileEventsMap& mq_fevts_map = mqFileEventsMap();
+  const int wrk_index = vppcom_worker_index();
+  RELEASE_ASSERT(wrk_index != -1, "");
+  if (mq_fevts_map.find(wrk_index) != mq_fevts_map.end()) {
+    return;
+  }
+  mq_fevts_map[wrk_index] = dispatcher.createFileEvent(
+      vppcom_mq_epoll_fd(), [](uint32_t events) -> void { onMqSocketEvents(events); },
+      Event::FileTriggerType::Edge, Event::FileReadyType::Read | Event::FileReadyType::Write);
+}
+
+void vclInterfaceInit(Event::Dispatcher& dispatcher, uint32_t concurrency) {
+  MqFileEventsMap& mq_fevts_map = mqFileEventsMap();
+  vppcom_app_create("envoy");
+  const int wrk_index = vppcom_worker_index();
+  std::vector<uint32_t>& epoll_handles = epollHandles();
+  // Assume we may have additional threads that request network access
+  epoll_handles.reserve(concurrency * 2);
+  epoll_handles[wrk_index] = vppcom_epoll_create();
```

Comment:
My bad, I missed this. Increasing a vector's capacity doesn't change its size. Then accessing a non-initialized element with `operator[]` is undefined behavior even though it should work ok in practice. Add actual `epoll_handles.resize()`.
Also `ASSERT` checking `wrk_index < epoll_handles.size()` wouldn't harm. If perf is not a concern then `operator[]` could be replaced with `at()` which checks ranges automatically.

Reply:
Done